scala-package/core/src/test/scala/ml/dmlc/mxnet/util/WarnIfNotDiposedSuite.scala

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ml.dmlc.mxnet
+
+import org.scalatest.{BeforeAndAfterAll, FunSuite}
+
+// scalastyle:off finalize
+class Leakable(enableTracing: Boolean = false, markDisposed: Boolean = false)
+    extends WarnIfNotDisposed {
+  def isDisposed: Boolean = markDisposed
+  override protected def tracingEnabled = enableTracing
+
+  var warningWasLogged: Boolean = false
```

Comment:
Consider using a type which represents more than two values in order to support the states "unitialized", "true" and "false".

Reply:
Which of these values do you think has these semantics? `isDisposed` and `tracingEnabled` only have set values, and `warningWasLogged` is a flag which is set to true at some point, I don't think a third value would make sense for any of them. Am I misunderstanding?

Reply:
In your tests you're checking based on ``warningWasLogged``, which is false by default. In line 55 you're checking if it's false, but you actually can't assure that your test actually succeeded as false is the default value.

Reply:
I see. I think the distinction here is that I think of this as an flag about "side-effect happened", and you're thinking about it as something which is set directly by the code under test (which, at present, it's not). Since the code under test is supposed to do _nothing_ in the case where the result is `false`, there's no mechanism for it to set the value to `false`.

This would be fixable by changing the code under test but I _think_ it'd make it less clear (and the change has been accepted anyway). Thanks very much for your comment though, much appreciate the feedback.